lib/usd/ui/views/TreeModel.cpp

```diff
@@ -0,0 +1,239 @@
+		//		  checkbox. But with context menus (to come) it might not.
+		const Qt::CheckState currState = item->checkState();
+		if (currState != state)
+		{
+			if (rMin == -1) rMin = r;
+			if (r > rMax) rMax = r;
+			item->setCheckState(state);
+			if (hasChildren(childIndex))
+				setChildCheckState(childIndex, state);
+		}
+	}
+	QModelIndex rMinIndex = this->index(rMin, 0, parent);
+	QModelIndex rMaxIndex = this->index(rMax, 0, parent);
+	emit dataChanged(rMinIndex, rMaxIndex);
+}
+
+void TreeModel::fillStagePopulationMask(UsdStagePopulationMask& popMask, const QModelIndex& parent)
+{
+	bool allChildrenChecked = true;		// Start with this assumption
+	for (int r=0; r<rowCount(parent); ++r)
+	{
+		QModelIndex childIndex = this->index(r, 0, parent);
+		TreeItem* item = static_cast<TreeItem*>(itemFromIndex(childIndex));
+		if (item->checkState() == Qt::Checked)
+		{
+			if (hasChildren(childIndex))
+			{
+				fillStagePopulationMask(popMask, childIndex);
+			}
+			auto primPath = item->prim().GetPath();
+			if (!popMask.Includes(primPath))
+				popMask.Add(primPath);
+		}
+		else
+			allChildrenChecked = false;
+	}
+	if (allChildrenChecked)
+	{
+		// We would have added each individual child item's paths to the pop mask.
+		// Instead let's add the parent path.
+		TreeItem* item = static_cast<TreeItem*>(itemFromIndex(parent));
+		if ((item != nullptr) && !item->prim().IsPseudoRoot())
+		{
+			auto primPath = item->prim().GetPath();
+			popMask.Add(primPath);
+		}
+	}
+}
+
+void TreeModel::fillPrimVariantSelections(ImportData::PrimVariantSelections& primVariantSelections, const QModelIndex& parent)
+{
+	for (int r=0; r<rowCount(parent); ++r)
+	{
+		QModelIndex variantIndex = this->index(r, kTreeColumn_Variants, parent);
+		TreeItem* item = static_cast<TreeItem*>(itemFromIndex(variantIndex));
+		if (item->variantSelectionModified())
+		{
+			// Note: both the variant name and variant selection roles contain a
+			//		 QStringList for data.
+			QVariant varSel = variantIndex.data(ItemDelegate::kVariantSelectionRole);
+			if (varSel.isValid() && varSel.canConvert<QStringList>())
+			{
+				// The name role must be valid if the selection role was.
+				QStringList varSelections = varSel.toStringList();
+				QStringList varNames = variantIndex.data(ItemDelegate::kVariantNameRole).toStringList();
+				assert(varSelections.count() == varNames.count());
+				if (varSelections.count() == varNames.count())
+				{
+					ImportData::VariantSelections varSels;
+					for (int i=0; i<varNames.count(); ++i)
+					{
+						varSels[varNames.at(i).toStdString()] = varSelections.at(i).toStdString();
+					}
+					primVariantSelections[item->prim().GetPath()] = varSels;
+				}
+			}
+		}
+
+		QModelIndex childIndex = this->index(r, 0, parent);
+		if (hasChildren(childIndex))
+		{
+			fillPrimVariantSelections(primVariantSelections, childIndex);
+		}
+	}
+}
+
+void TreeModel::openPersistentEditors(QTreeView* tv, const QModelIndex& parent)
```

Comment:
Helper called after tree is created to create all the variant set widgets on each row (that has variant sets).

In Qt terms this is called a persistent editor. Meaning it's always there, rather than only being created when you click in a cell.